File: src/coral/common/hash.zig
const std = @import("std");

pub const HashAlgorithm = enum {
    sha256,
    sha512,
    blake3,

    pub fn digestLength(self: HashAlgorithm) usize {
        return switch (self) {
            .sha256 => 32,
            .sha512 => 64,
            .blake3 => 32,
        };
    }
};

pub fn hashFile(
    allocator: std.mem.Allocator,
    path: []const u8,
    algorithm: HashAlgorithm,
) ![]const u8 {
    const file = try std.fs.cwd().openFile(path, .{});
    defer file.close();

    var buf: [65536]u8 = undefined;

    switch (algorithm) {
        .sha256 => {
            var hasher = std.crypto.hash.sha2.Sha256.init(.{});
            while (true) {
                const n = file.read(&buf) catch break;
                if (n == 0) break;
                hasher.update(buf[0..n]);
            }
            var out: [32]u8 = undefined;
            hasher.final(&out);
            const hex = std.fmt.bytesToHex(out, .lower);
            return allocator.dupe(u8, &hex);
        },
        .sha512 => {
            var hasher = std.crypto.hash.sha2.Sha512.init(.{});
            while (true) {
                const n = file.read(&buf) catch break;
                if (n == 0) break;
                hasher.update(buf[0..n]);
            }
            var out: [64]u8 = undefined;
            hasher.final(&out);
            const hex = std.fmt.bytesToHex(out, .lower);
            return allocator.dupe(u8, &hex);
        },
        .blake3 => {
            var hasher = std.crypto.hash.Blake3.init(.{});
            while (true) {
                const n = file.read(&buf) catch break;
                if (n == 0) break;
                hasher.update(buf[0..n]);
            }
            var out: [32]u8 = undefined;
            hasher.final(&out);
            const hex = std.fmt.bytesToHex(out, .lower);
            return allocator.dupe(u8, &hex);
        },
    }
}

pub const BatchHashResult = struct {
    path: []const u8,
    hash: ?[]const u8,
    error_msg: ?[]const u8,
};

pub fn hashBatch(
    allocator: std.mem.Allocator,
    paths: []const []const u8,
    algorithm: HashAlgorithm,
    progress_callback: ?*const fn (usize, usize) void,
) ![]BatchHashResult {
    var results = try allocator.alloc(BatchHashResult, paths.len);
    errdefer allocator.free(results);

    for (paths, 0..) |path, i| {
        if (progress_callback) |cb| cb(i, paths.len);

        results[i] = .{ .path = path, .hash = null, .error_msg = null };

        const hash = hashFile(allocator, path, algorithm) catch |err| {
            results[i].error_msg = try std.fmt.allocPrint(allocator, "{}", .{err});
            continue;
        };
        results[i].hash = hash;
    }

    return results;
}

pub fn hashString(
    allocator: std.mem.Allocator,
    data: []const u8,
    algorithm: HashAlgorithm,
) ![]const u8 {
    switch (algorithm) {
        .sha256 => {
            var out: [32]u8 = undefined;
            std.crypto.hash.sha2.Sha256.hash(data, &out, .{});
            const hex = std.fmt.bytesToHex(out, .lower);
            return allocator.dupe(u8, &hex);
        },
        .sha512 => {
            var out: [64]u8 = undefined;
            std.crypto.hash.sha2.Sha512.hash(data, &out, .{});
            const hex = std.fmt.bytesToHex(out, .lower);
            return allocator.dupe(u8, &hex);
        },
        .blake3 => {
            var out: [32]u8 = undefined;
            std.crypto.hash.Blake3.hash(data, &out, .{});
            const hex = std.fmt.bytesToHex(out, .lower);
            return allocator.dupe(u8, &hex);
        },
    }
}

pub fn blake3Hash(data: []const u8) [32]u8 {
    var out: [32]u8 = undefined;
    std.crypto.hash.Blake3.hash(data, &out, .{});
    return out;
}

pub fn blake3Hex(allocator: std.mem.Allocator, data: []const u8) ![]const u8 {
    const h = blake3Hash(data);
    const hex = std.fmt.bytesToHex(h, .lower);
    return allocator.dupe(u8, &hex);
}

pub const HashState = struct {
    algorithm: HashAlgorithm,
    sha256: ?std.crypto.hash.sha2.Sha256 = null,
    sha512: ?std.crypto.hash.sha2.Sha512 = null,
    blake3_h: ?std.crypto.hash.Blake3 = null,

    pub fn init(algorithm: HashAlgorithm) HashState {
        return switch (algorithm) {
            .sha256 => .{ .algorithm = algorithm, .sha256 = std.crypto.hash.sha2.Sha256.init(.{}) },
            .sha512 => .{ .algorithm = algorithm, .sha512 = std.crypto.hash.sha2.Sha512.init(.{}) },
            .blake3 => .{ .algorithm = algorithm, .blake3_h = std.crypto.hash.Blake3.init(.{}) },
        };
    }

    pub fn update(self: *HashState, data: []const u8) void {
        switch (self.algorithm) {
            .sha256 => self.sha256.?.update(data),
            .sha512 => self.sha512.?.update(data),
            .blake3 => self.blake3_h.?.update(data),
        }
    }

    pub fn final(self: *HashState, allocator: std.mem.Allocator) ![]const u8 {
        switch (self.algorithm) {
            .sha256 => {
                var out: [32]u8 = undefined;
                self.sha256.?.final(&out);
                const hex = std.fmt.bytesToHex(out, .lower);
                return allocator.dupe(u8, &hex);
            },
            .sha512 => {
                var out: [64]u8 = undefined;
                self.sha512.?.final(&out);
                const hex = std.fmt.bytesToHex(out, .lower);
                return allocator.dupe(u8, &hex);
            },
            .blake3 => {
                var out: [32]u8 = undefined;
                self.blake3_h.?.final(&out);
                const hex = std.fmt.bytesToHex(out, .lower);
                return allocator.dupe(u8, &hex);
            },
        }
    }
};

const testing = std.testing;

test "hashString: sha256" {
    const result = try hashString(testing.allocator, "hello", .sha256);
    defer testing.allocator.free(result);
    try testing.expectEqual(@as(usize, 64), result.len);
}

test "hashString: blake3" {
    const result = try hashString(testing.allocator, "hello", .blake3);
    defer testing.allocator.free(result);
    try testing.expectEqual(@as(usize, 64), result.len);
}

test "blake3Hash: returns 32 bytes" {
    const h = blake3Hash("hello");
    try testing.expectEqual(@as(usize, 32), h.len);
}

test "HashState: incremental hashing matches single-pass" {
    var state = HashState.init(.sha256);
    state.update("hello");
    state.update(" ");
    state.update("world");
    const result = try state.final(testing.allocator);
    defer testing.allocator.free(result);

    var single = HashState.init(.sha256);
    single.update("hello world");
    const expected = try single.final(testing.allocator);
    defer testing.allocator.free(expected);

    try testing.expectEqualStrings(expected, result);
}

test "HashAlgorithm: digestLength" {
    try testing.expectEqual(@as(usize, 32), HashAlgorithm.sha256.digestLength());
    try testing.expectEqual(@as(usize, 64), HashAlgorithm.sha512.digestLength());
    try testing.expectEqual(@as(usize, 32), HashAlgorithm.blake3.digestLength());
}

File: src/common-core/src/hash.rs
//! Hashing utilities: `blake3_*`, `sha256_*`, `fnv1a64`, `HashState`, `hash_file`/`hash_batch`.

use blake3::Hasher;
use sha2::{Digest, Sha256, Sha512};
use std::io::Read;
use std::path::Path;

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum HashAlgorithm {
    Sha256,
    Sha512,
    Blake3,
}

impl HashAlgorithm {
    pub fn digest_length(self) -> usize {
        match self {
            HashAlgorithm::Sha512 => 64,
            HashAlgorithm::Sha256 | HashAlgorithm::Blake3 => 32,
        }
    }
}

pub fn sha256_hex(data: &[u8]) -> String {
    let mut hasher = Sha256::new();
    hasher.update(data);
    let result = hasher.finalize();
    hex_encode(&result)
}

/// sha256 hex of a string's bytes (file-id domain: ids hash path strings,
/// not file contents).
///
/// # Examples
///
/// ```
/// use common_core::hash::sha256_str;
///
/// assert_eq!(
///     sha256_str("hello"),
///     "2cf24dba5fb0a30e26e83b2ac5b9e29e1b161e5c1fa7425e73043362938b9824"
/// );
/// ```
#[must_use]
pub fn sha256_str(data: &str) -> String {
    sha256_hex(data.as_bytes())
}

/// sha256 hex of a file's bytes; `None` when the file cannot be read
/// (fail-open: callers treat unreadable files as changed, never as
/// current). Streams through the shared [`hash_file`] engine, so the
/// digest is byte-identical to [`sha256_hex`] over the full contents.
///
/// # Examples
///
/// ```no_run
/// use std::path::Path;
/// use common_core::hash::sha256_file;
///
/// let hash = sha256_file(Path::new("Cargo.toml"));
/// assert!(hash.is_some());
/// ```
#[must_use]
pub fn sha256_file(path: &Path) -> Option<String> {
    hash_file(path, HashAlgorithm::Sha256).ok()
}

pub fn blake3_hash(data: &[u8]) -> [u8; 32] {
    blake3::hash(data).into()
}

/// Compute a BLAKE3 content hash and return it as a lowercase hex string.
///
/// # Examples
///
/// ```
/// use common_core::hash::blake3_hex;
///
/// let hash = blake3_hex(b"hello world");
/// assert_eq!(hash.len(), 64); // 32 bytes = 64 hex chars
/// assert!(!hash.is_empty());
/// ```
pub fn blake3_hex(data: &[u8]) -> String {
    blake3::hash(data).to_hex().to_string()
}

pub fn fnv1a64(input: &[u8]) -> u64 {
    const OFFSET: u64 = 14_695_981_039_346_656_037;
    const PRIME: u64 = 1_099_511_628_211;
    let mut hash: u64 = OFFSET;
    for &byte in input {
        hash ^= u64::from(byte);
        hash = hash.wrapping_mul(PRIME);
    }
    hash
}

pub fn content_hash_with_model(content: &str, model: &str) -> [u8; 16] {
    let mut hasher = Sha256::new();
    hasher.update(model.as_bytes());
    hasher.update(b"\x00");
    hasher.update(content.as_bytes());
    let result = hasher.finalize();
    let mut out = [0u8; 16];
    out.copy_from_slice(&result[..16]);
    out
}

enum HashStateInner {
    Sha256(Sha256),
    Sha512(Sha512),
    Blake3(Box<Hasher>),
}

pub struct HashState {
    inner: HashStateInner,
}

impl HashState {
    pub fn new(algorithm: HashAlgorithm) -> Self {
        match algorithm {
            HashAlgorithm::Sha256 => Self {
                inner: HashStateInner::Sha256(Sha256::new()),
            },
            HashAlgorithm::Sha512 => Self {
                inner: HashStateInner::Sha512(Sha512::new()),
            },
            HashAlgorithm::Blake3 => Self {
                inner: HashStateInner::Blake3(Box::new(Hasher::new())),
            },
        }
    }

    pub fn update(&mut self, data: &[u8]) {
        match &mut self.inner {
            HashStateInner::Sha256(h) => h.update(data),
            HashStateInner::Sha512(h) => h.update(data),
            HashStateInner::Blake3(h) => {
                h.update(data);
            }
        }
    }

    pub fn finalize(&self) -> Vec<u8> {
        match &self.inner {
            HashStateInner::Sha256(h) => h.clone().finalize().to_vec(),
            HashStateInner::Sha512(h) => h.clone().finalize().to_vec(),
            HashStateInner::Blake3(h) => h.clone().finalize().as_bytes().to_vec(),
        }
    }

    pub fn digest_hex(&self) -> String {
        hex_encode(&self.finalize())
    }
}

#[derive(Debug, Clone)]
pub struct BatchHashResult {
    pub path: String,
    pub hash: Option<String>,
}

pub fn hash_file(path: &Path, algorithm: HashAlgorithm) -> std::io::Result<String> {
    let mut file = std::fs::File::open(path)?;
    let mut hasher = HashState::new(algorithm);
    let mut buf = [0u8; 65536];
    loop {
        let n = file.read(&mut buf)?;
        if n == 0 {
            break;
        }
        hasher.update(&buf[..n]);
    }
    Ok(hasher.digest_hex())
}

pub fn hash_batch(paths: &[std::path::PathBuf], algorithm: HashAlgorithm) -> Vec<BatchHashResult> {
    paths
        .iter()
        .map(|path| {
            let hash = hash_file(path, algorithm).ok();
            BatchHashResult {
                path: path.to_string_lossy().to_string(),
                hash,
            }
        })
        .collect()
}

pub fn sha256_digest(data: &[u8]) -> [u8; 32] {
    let mut hasher = Sha256::new();
    hasher.update(data);
    let result = hasher.finalize();
    let mut out = [0u8; 32];
    out.copy_from_slice(&result);
    out
}

pub fn hex_encode(data: &[u8]) -> String {
    data.iter()
        .map(|b| format!("{b:02x}"))
        .collect::<Vec<_>>()
        .concat()
}

/// Generate a time-based pseudo-UUID v4 string.
///
/// Uses `SystemTime` nanoseconds, process ID, and `fastrand` for entropy.
/// Formatted as `xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx` (36 chars, 4 dashes).
pub fn uuid_v4() -> String {
    use std::time::{SystemTime, UNIX_EPOCH};
    let now = SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .unwrap_or_default();
    let nanos = now.as_nanos();
    let pid = u64::from(std::process::id());
    let rng = fastrand::u64(..);
    let time_hi = (nanos >> 32) as u32;
    let time_mid = (nanos >> 16) as u32 & 0xFFFF;
    let time_lo = nanos as u32 & 0xFFFF;
    let version = (u64::from(pid as u32 & 0xFFF)) as u32 | 0x4000;
    let variant = (((pid >> 12) ^ rng) as u32 & 0x3FFF) | 0x8000;
    let node = (rng >> 32) as u32;
    format!("{time_hi:08x}-{time_mid:04x}-{time_lo:04x}-{version:04x}-{variant:04x}{node:08x}")
}

